File: make_dendrogram_bar.py
"""
Created on Sat Apr 3 18:39:34 2021
Last modified on Thurs March 31 09:16:00 2022

@author: 
Original matlab code for graph filtration: Hye Kyung Lee
Translation to python and addition of visualization code: Sungwoo Bae with the help of matlab2python

"""
import numpy as np
    
def make_dendrogram_bar(history,duration,cvertical_x=None,cvertical_y=None,chorizontal_x=None,chorizontal_y=None,cdots=None): 
    if (cvertical_x is None) and (cvertical_y is None) and (chorizontal_x is None) \
            and (chorizontal_y is None) and (cdots is None):
        is_new = 1
    else:
        is_new = 0
    
    ncc = duration.shape[0]

    # Estimate the depth of dendrogram
    nlayer = []
    # Find CCs with no parent
    length_history = np.array(list(map(lambda x: len(x), history)))
    ind_notempty = np.where(np.sum(duration, axis=1) != 0)[0]
    ind_empty = np.setdiff1d([range(len(history))], ind_notempty)
    ind_past = np.setdiff1d(np.where(length_history == 0)[0], ind_empty)
    nlayer.append(ind_past.tolist())

    while len(ind_past) < len(ind_notempty):
        tind = np.array(list(map(lambda x: len(np.intersect1d(x, ind_past)) == len(x), history)))
        ttind = np.setdiff1d(np.where(tind)[0], np.concatenate((ind_past,ind_empty)))
        if len(ttind) != 0:
            nlayer.append(ttind.tolist())
            ind_past = np.concatenate((ind_past, ttind))
    
    if is_new == 1:
        # Estimate bars in a dendrogram
        nvertical_x = np.zeros((ncc,2))
        nvertical_y = np.zeros((ncc,2))
        nhorizontal_x = np.zeros((ncc,2))
        nhorizontal_y = np.zeros((ncc,2))
        ndots = np.zeros((ncc,2))
        
        sval_ind = dict(sorted(zip(range(len(duration[nlayer[0],1])), duration[nlayer[0],1]), key=lambda x: x[1], reverse=True))
        sval = np.array(list(sval_ind.values()))
        sind = np.array(list(sval_ind.keys())).astype(int)
        sind = np.array(nlayer[0])[sind]

        for i in range(len(sind)):
            ii = sind[i]
            nvertical_x[ii,:] = np.array([i,i])
            nvertical_y[ii,:] = np.array([duration[ii,0], duration[ii,1]])
            ndots[ii,:] = np.array([i, duration[ii,0]])

        for i in range(1,len(nlayer)):
            for j in range(len(nlayer[i])):
                tx = nvertical_x[history[nlayer[i][j]], 0]
                if len(tx)>0: 
                    nvertical_x[nlayer[i][j],:] = np.mean(tx) * np.ones((1,2))
                    nhorizontal_x[nlayer[i][j],:] = [np.min(tx), np.max(tx)]
                    ndots[nlayer[i][j],0] = np.mean(tx)
                ndots[nlayer[i][j],1] = duration[nlayer[i][j],0]
                nvertical_y[nlayer[i][j],:] = duration[nlayer[i][j],:]
                nhorizontal_y[nlayer[i][j],:] = duration[nlayer[i][j],0] * np.ones((1,2))

    else:
        ncc = duration.shape[0]

        nvertical_x = cvertical_x
        nvertical_y = cvertical_y
        nhorizontal_x = chorizontal_x
        nhorizontal_y = chorizontal_y
        ndots = cdots

        nvertical_x[ind_empty,:] = 0
        nvertical_y[ind_empty,:] = 0
        nhorizontal_x[ind_empty,:] = 0
        nhorizontal_y[ind_empty,:] = 0
        ndots[ind_empty,:] = 0

        for j in range(len(nlayer[0])):
            ii = nlayer[0][j]
            nvertical_y[ii,:] = np.sort(duration[ii,:])
            nhorizontal_x[ii,:] = 0
            nhorizontal_y[ii,:] = 0
            ndots[ii,:] = np.array([nvertical_x[ii,0], nvertical_y[ii,1]])
        
        for i in range(1,len(nlayer)):
            for j in range(len(nlayer[i])):
                ii = nlayer[i][j]
                tx = nvertical_x[history[ii],0]
                if len(tx)>0:
                    nvertical_x[ii,:] = np.mean(tx) * np.ones((1,2))
                    nhorizontal_x[ii,:] = [np.min(tx), np.max(tx)]
                    ndots[ii,0] = np.mean(tx)
                ndots[ii,1] = duration[ii,0]
                nvertical_y[ii,:] = duration[ii,:]
                nhorizontal_y[ii,:] = duration[ii,0] * np.ones((1,2))
    
    return nvertical_x,nvertical_y,nhorizontal_x,nhorizontal_y,ndots,nlayer

File: src/make_dendrogram_bar.cpp
#include "make_dendrogram_bar.h"
#include <algorithm>
#include <numeric>
#include <set>
#include <limits>
#include <cmath>      // For std::isnan
#include <iostream>   // For debugging output (optional)

std::tuple<
    Eigen::MatrixXd,
    Eigen::MatrixXd,
    Eigen::MatrixXd,
    Eigen::MatrixXd,
    Eigen::MatrixXd,
    std::vector<std::vector<int>>
>
make_dendrogram_bar(
    const std::vector<std::vector<int>>& history,
    const Eigen::MatrixXd& duration,
    const Eigen::MatrixXd& cvertical_x,
    const Eigen::MatrixXd& cvertical_y,
    const Eigen::MatrixXd& chorizontal_x,
    const Eigen::MatrixXd& chorizontal_y,
    const Eigen::MatrixXd& cdots
)
{
    bool is_new = (cvertical_x.size() == 0) && (cvertical_y.size() == 0) &&
                  (chorizontal_x.size() == 0) && (chorizontal_y.size() == 0) &&
                  (cdots.size() == 0);

    int ncc = duration.rows();

    // Estimate the depth of dendrogram
    std::vector<std::vector<int>> nlayer;

    // Find CCs with no parent
    std::vector<int> length_history(ncc);
    for (int i = 0; i < ncc; ++i) {
        length_history[i] = history[i].size();
    }

    // Identify non-empty and empty indices based on the validity of duration values
    std::vector<int> ind_notempty;
    std::vector<int> ind_empty;
    const double EPSILON = 0; // Tolerance for floating-point comparison
    for (int i = 0; i < ncc; ++i) {
        double row_sum = duration.row(i).sum();
        if (std::abs(row_sum) > EPSILON) {
            ind_notempty.push_back(i);
        } else {
            ind_empty.push_back(i);
        }
    }

    // Leaf CCs
    std::vector<int> ind_past;
    for (int i = 0; i < ncc; ++i) {
        if (length_history[i] == 0 &&
            std::find(ind_empty.begin(), ind_empty.end(), i) == ind_empty.end()) {
            ind_past.push_back(i);
        }
    }
    nlayer.push_back(ind_past);

    // Build layers
    while (static_cast<int>(ind_past.size()) < static_cast<int>(ind_notempty.size())) {
        std::vector<bool> tind(ncc, false);
        for (int i = 0; i < ncc; ++i) {
            if (!history[i].empty()) {
                std::vector<int> intersect;
                std::set_intersection(
                    history[i].begin(), history[i].end(),
                    ind_past.begin(), ind_past.end(),
                    std::back_inserter(intersect)
                );
                tind[i] = (intersect.size() == history[i].size());
            }
        }

        std::vector<int> ttind;
        for (int i = 0; i < ncc; ++i) {
            if (tind[i] &&
                std::find(ind_past.begin(), ind_past.end(), i) == ind_past.end() &&
                std::find(ind_empty.begin(), ind_empty.end(), i) == ind_empty.end()) {
                ttind.push_back(i);
            }
        }

        if (!ttind.empty()) {
            nlayer.push_back(ttind);
            ind_past.insert(ind_past.end(), ttind.begin(), ttind.end());
            // Sort ind_past to ensure set_intersection works correctly
            std::sort(ind_past.begin(), ind_past.end());
        } else {
            break;
        }
    }

    Eigen::MatrixXd nvertical_x;
    Eigen::MatrixXd nvertical_y;
    Eigen::MatrixXd nhorizontal_x;
    Eigen::MatrixXd nhorizontal_y;
    Eigen::MatrixXd ndots;

    if (is_new) {
        nvertical_x = Eigen::MatrixXd::Zero(ncc, 2);
        nvertical_y = Eigen::MatrixXd::Zero(ncc, 2);
        nhorizontal_x = Eigen::MatrixXd::Zero(ncc, 2);
        nhorizontal_y = Eigen::MatrixXd::Zero(ncc, 2);
        ndots = Eigen::MatrixXd::Zero(ncc, 2);

        // Sort nlayer[0] by duration
        std::vector<std::pair<int, double>> sval_ind;
        for (int idx : nlayer[0]) {
            sval_ind.emplace_back(idx, duration(idx, 1));
        }

        // Stable sort to handle ties
        std::stable_sort(sval_ind.begin(), sval_ind.end(),
                         [](const std::pair<int, double>& a, const std::pair<int, double>& b) {
                             return a.second > b.second;
                         });

        std::vector<int> sind;
        for (const auto& pair : sval_ind) {
            sind.push_back(pair.first);
        }

        for (size_t i = 0; i < sind.size(); ++i) {
            int ii = sind[i];
            nvertical_x.row(ii) = Eigen::Vector2d(i, i);
            nvertical_y.row(ii) = duration.row(ii);
            ndots.row(ii) = Eigen::Vector2d(i, duration(ii, 0));
        }

        for (size_t layer_idx = 1; layer_idx < nlayer.size(); ++layer_idx) {
            for (int idx : nlayer[layer_idx]) {
                std::vector<double> tx;
                for (int h : history[idx]) {
                    tx.push_back(nvertical_x(h, 0));
                }
                if (!tx.empty()) {
                    // Compute mean, min, max of tx
                    double sum_tx = std::accumulate(tx.begin(), tx.end(), 0.0);
                    double mean_tx = sum_tx / tx.size();
                    double min_tx = *std::min_element(tx.begin(), tx.end());
                    double max_tx = *std::max_element(tx.begin(), tx.end());

                    nvertical_x.row(idx) = Eigen::Vector2d(mean_tx, mean_tx);
                    nhorizontal_x.row(idx) = Eigen::Vector2d(min_tx, max_tx);
                    ndots(idx, 0) = mean_tx;
                }
                ndots(idx, 1) = duration(idx, 0);
                nvertical_y.row(idx) = duration.row(idx);
                nhorizontal_y.row(idx) = Eigen::Vector2d(duration(idx, 0), duration(idx, 0));
            }
        }
    } else {
        // Use existing matrices
        nvertical_x = cvertical_x;
        nvertical_y = cvertical_y;
        nhorizontal_x = chorizontal_x;
        nhorizontal_y = chorizontal_y;
        ndots = cdots;

        // Remove resizing if input matrices are guaranteed to be (ncc, 2)
        // If not, ensure that input matrices are correctly sized before passing to this function

        // Set ind_empty rows to zero
        for (int idx : ind_empty) {
            nvertical_x.row(idx).setZero();
            nvertical_y.row(idx).setZero();
            nhorizontal_x.row(idx).setZero();
            nhorizontal_y.row(idx).setZero();
            ndots.row(idx).setZero();
        }

        // Update leaf nodes
        for (int ii : nlayer[0]) {
            // Sort duration.row(ii)
            Eigen::RowVector2d sorted_duration = duration.row(ii);
            if (sorted_duration(0) > sorted_duration(1)) {
                std::swap(sorted_duration(0), sorted_duration(1));
            }
            nvertical_y.row(ii) = sorted_duration;

            // Set nhorizontal_x and nhorizontal_y to zero
            nhorizontal_x.row(ii).setZero();
            nhorizontal_y.row(ii).setZero();

            // Assign ndots
            ndots.row(ii) = Eigen::RowVector2d(nvertical_x(ii, 0), nvertical_y(ii, 1));
        }

        // Process subsequent layers
        for (size_t layer_idx = 1; layer_idx < nlayer.size(); ++layer_idx) {
            for (int idx : nlayer[layer_idx]) {
                std::vector<double> tx;
                for (int h : history[idx]) {
                    tx.push_back(nvertical_x(h, 0));
                }
                if (!tx.empty()) {
                    double sum_tx = std::accumulate(tx.begin(), tx.end(), 0.0);
                    double mean_tx = sum_tx / tx.size();
                    double min_tx = *std::min_element(tx.begin(), tx.end());
                    double max_tx = *std::max_element(tx.begin(), tx.end());

                    nvertical_x.row(idx) = Eigen::RowVector2d(mean_tx, mean_tx);
                    nhorizontal_x.row(idx) = Eigen::RowVector2d(min_tx, max_tx);
                    ndots(idx, 0) = mean_tx;
                }
                ndots(idx, 1) = duration(idx, 0);
                nvertical_y.row(idx) = duration.row(idx);
                nhorizontal_y.row(idx) = Eigen::RowVector2d(duration(idx, 0), duration(idx, 0));
            }
        }
    }

    return std::make_tuple(nvertical_x, nvertical_y, nhorizontal_x, nhorizontal_y, ndots, nlayer);
}